### PR TITLE
Refactoring WadRayMath library

### PR DESCRIPTION
**Pull request by kaliberpoziomka**
This PR is proposing the refactoring of WadRayMath. Refactoring includes:
- wrapping library in namespace
- removing Wad and Ray types, instead, functions are working on Uint256 types

Removing types is proposed since it is consistent with the Solidity implementation of WadRayMath and allows writing cleaner code because reduces the number of conversions required by the current implementation.
The disadvantage is that because of the lack of types in functions, code is more prone to errors - e.g. rayMul(a, b) where a and b are Wad numbers would result in a number that is of magnitude 1e9.
Because of those issues, this PR is a field of discussion.

NOTE: Other contracts/libraries which are currently using WadRayMath are refactored in this PR so that errors won't occur.

### CODE REVIEW

**Comment by hsouf** [comment]
Thank you @kaliberpoziomka for this refactoring! I'll be happy if the ci is green. Also, can we add in this PR the wad ray tests we've got [here](https://github.com/aave-starknet-project/aave-starknet-bridge/blob/main/contracts/l2/test/wad_ray_tests.cairo)(from phase 1) after we migrate them to the new refactoring?

File: contracts/protocol/libraries/math/wad_ray_math.cairo [hunk 1]
```diff
@@ -10,244 +10,238 @@ from starkware.cairo.common.uint256 import (
...
-        return (Wad(Uint256(0, 0)))
+
+    func uint256_max() -> (max : Uint256):
+        return (Uint256(UINT128_MAX, UINT128_MAX))
     end
 
-    let (UINT256_MAX) = uint256_max()
-    let (HALF_WAD_UINT) = half_wad()
-    let (WAD_UINT) = wad()
+    func wad_mul{range_check_ptr}(a : Uint256, b : Uint256) -> (res : Uint256):
```
Comment (achab) [inline contracts/protocol/libraries/math/wad_ray_math.cairo:54]: it's a good time to add a comment to explain the behavior of the function

File: contracts/protocol/libraries/math/wad_ray_math.cairo [hunk 2]
```diff
@@ -10,244 +10,238 @@ from starkware.cairo.common.uint256 import (
...
-    let (ab, _) = uint256_mul(a.wad, b.wad)
-    let (abHW, _) = uint256_add(ab, HALF_WAD_UINT.wad)
-    let (res, _) = uint256_unsigned_div_rem(abHW, WAD_UINT.wad)
-    return (Wad(res))
-end
+        with_attr error_message("WAD multiplication overflow"):
+            let (bound) = uint256_sub(UINT256_MAX, HALF_WAD_UINT)
+            let (quotient, rem) = uint256_unsigned_div_rem(bound, b)
+            let (le) = uint256_le(a, quotient)
+            assert le = 1
```
Comment (achab) [inline contracts/protocol/libraries/math/wad_ray_math.cairo:71]: `assert le = TRUE`

File: contracts/protocol/libraries/math/wad_ray_math.cairo [hunk 3]
```diff
@@ -10,244 +10,238 @@ from starkware.cairo.common.uint256 import (
...
-    let (aWAD, _) = uint256_mul(a.wad, WAD_UINT.wad)
-    let (aWADHalfB, _) = uint256_add(aWAD, halfB)
-    let (res, _) = uint256_unsigned_div_rem(aWADHalfB, b.wad)
-    return (Wad(res))
-end
+        with_attr error_message("WAD div overflow"):
+            let (bound) = uint256_sub(UINT256_MAX, halfB)
+            let (quo, _) = uint256_unsigned_div_rem(bound, WAD_UINT)
+            let (le) = uint256_le(a, quo)
+            assert le = 1
```
Comment (achab) [inline contracts/protocol/libraries/math/wad_ray_math.cairo:95]: `assert le = TRUE`

File: contracts/protocol/libraries/math/wad_ray_math.cairo [hunk 4]
```diff
@@ -10,244 +10,238 @@ from starkware.cairo.common.uint256 import (
...
-    let (ab, _) = uint256_mul(a.ray, b.ray)
-    let (abHR, _) = uint256_add(ab, HALF_RAY_UINT.ray)
-    let (res, _) = uint256_unsigned_div_rem(abHR, RAY_UINT.ray)
-    return (Ray(res))
-end
+        with_attr error_message("RAY div overflow"):
+            let (bound) = uint256_sub(UINT256_MAX, HALF_RAY_UINT)
+            let (quotient, rem) = uint256_unsigned_div_rem(bound, b)
+            let (le) = uint256_le(a, quotient)
+            assert le = 1
```
Comment (achab) [inline contracts/protocol/libraries/math/wad_ray_math.cairo:131]: same here

File: contracts/protocol/libraries/math/wad_ray_math.cairo [hunk 5]
```diff
@@ -10,244 +10,238 @@ from starkware.cairo.common.uint256 import (
...
+        alloc_locals
+        let (HALF_WAD_RAY_RATIO_UINT) = half_wad_ray_ratio()
+        let (WAD_RAY_RATIO_UINT) = wad_ray_ratio()
 
-    let (res, overflow) = uint256_mul(a.wad, WAD_RAY_RATIO_UINT)
-    with_attr error_message("wad_to_ray overflow"):
-        assert overflow.high + overflow.low = 0
+        let (res, overflow) = uint256_add(a, HALF_WAD_RAY_RATIO_UINT)
+        with_attr error_message("ray_to_wad overflow"):
+            assert overflow = 0
```
Comment (achab) [inline contracts/protocol/libraries/math/wad_ray_math.cairo:171]: `assert overflow = FALSE`

File: contracts/protocol/libraries/math/wad_ray_math.cairo [hunk 6]
```diff
@@ -10,244 +10,238 @@ from starkware.cairo.common.uint256 import (
...
+        return (diff)
+    end
 
-func wad_le{range_check_ptr}(a : Wad, b : Wad) -> (res : felt):
-    let a_wad = a.wad
-    let b_wad = b.wad
-    let (res) = uint256_le(a_wad, b_wad)
-    return (res)
+    func wad_le{range_check_ptr}(a : Uint256, b : Uint256) -> (res : felt):
+        let a_wad = a
```
Comment (achab) [inline contracts/protocol/libraries/math/wad_ray_math.cairo:242]: this line and the following one can be removed, and simply use `a` and `b` on line 244

**Review by achab — APPROVED** [review]
Thanks a lot for this @kaliberpoziomka ! Once these tiny comments are addressed, feel free to merge the PR. I would suggest @MakC-Ukr adapt his PR about tests on this once is merged.

**Comment by kaliberpoziomka** [comment]
@hsouf thanks for reminding me about P1. Shouldn't we create this PR in the P1 repo?

**Comment by hsouf** [comment]
> @hsouf thanks for reminding me about P1. Shouldn't we create this PR in the P1 repo?
No need, we will have to stick with current implementation on p1

**Comment by achab** [comment]
> Thank you @kaliberpoziomka for this refactoring! I'll be happy if the ci is green. Also, can we add in this PR the wad ray tests we've got [here](https://github.com/aave-starknet-project/aave-starknet-bridge/blob/main/contracts/l2/test/wad_ray_tests.cairo)(from phase 1) after we migrate them to the new refactoring?

@hsouf @kaliberpoziomka no need to add the tests in this PR since @MakC-Ukr is also working on it. btw @MakC-Ukr, you'll need to adapt a bit your PR once this one is merged.

**Review by achab — APPROVED** [review]

**Review by hsouf — APPROVED** [review]
Thank you @kaliberpoziomka for this PR👍🏻 let's ship it once ci is green.

**Pull request merged**